CircleCI-Public/circleci-yaml-language-server
# chore: Add rollbar

[Jira](https://circleci.atlassian.net/browse/DEVEX-1138)

**To the reviewer**: would you want to merge this PR after approving - which normally should not happen - please restrain yourself

# Description

This PR adds Rollbar on the LSP.
This PR is tightly coupled to a PR on the extension that makes sure the LSP receives the information it expects: https://github.com/circleci/circleci-vscode-extension/pull/799
It is important to note that because the VSCode extension must be updated to enable Rollbar on the LSP, people won't have Rollbar enabled on their Rollbar as long as their VSCode extension is not up to date.

# Implementation details

Because the logic of the LSP is contained in the [server.commandHandler function](https://github.com/CircleCI-Public/circleci-yaml-language-server/blob/92cca700613def609af1f26096c7022f8c351d65/pkg/server/server.go#L27-L78) and that this function runs in its own goroutine, it is not possible to simply wrap the call to the `main` function with `rollbar.WrapAndWait` as [it is first proposed in the documentation](https://pkg.go.dev/github.com/rollbar/rollbar-go#hdr-Basic_Usage). Wrapping `server.commandHandler` was not possible either because we would loose the return of the function. Instead the implementation is simply, [as advised in the doc](https://pkg.go.dev/github.com/rollbar/rollbar-go#hdr-Basic_Usage), to recover the panic in the function and to pass to panic to rollbar.
After recovering and sending the panic to rollbar, we still dispatch the panic again because the panic most likely stopped the execution for a reason and it's the client's responsability to relaunch the server.

# How to validate

You can add a `panic` in the LSP code, ask me for the rollbar token, build the LSP with this token, run the extension, open a YAML, trigger the `panic` you added and make sure your panic can be seen in Rollbar
